source/common/stats/stats_impl.cc

```diff
@@ -57,6 +61,55 @@ std::string Utility::sanitizeStatsName(const std::string& name) {
   return stats_name;
 }
 
+TagExtractorImpl::TagExtractorImpl(const std::string& name, const std::string& regex)
+    : name_(name), regex_(regex) {}
+
+TagExtractorPtr TagExtractorImpl::createTagExtractor(const std::string& name,
+                                                     const std::string& regex) {
+
+  if (name.empty()) {
+    throw EnvoyException("tag_name cannot be empty");
+  }
+
+  if (!regex.empty()) {
+    return TagExtractorPtr{new TagExtractorImpl(name, regex)};
+  } else {
+    // Look up the default for that name.
+    const auto& name_regex_pairs = Config::TagNames::get().name_regex_pairs_;
+    auto it = std::find_if(name_regex_pairs.begin(), name_regex_pairs.end(),
+                           [&name](const std::pair<std::string, std::string>& name_regex_pair) {
+                             return name == name_regex_pair.first;
+                           });
+    if (it != name_regex_pairs.end()) {
+      return TagExtractorPtr{new TagExtractorImpl(name, it->second)};
+    } else {
+      throw EnvoyException(fmt::format(
+          "No regex specified for tag specifier and no default regex for name: '{}'", name));
+    }
+  }
+}
+
+std::string TagExtractorImpl::extractTag(const std::string& tag_extracted_name,
+                                         std::vector<Tag>& tags) const {
+  std::smatch match;
+  // The regex must match and contain one or more subexpressions (all after the first are ignored).
```

Comment:
Maybe explain the `remove_subexpr` vs. `value_expr` distinction as well (i.e. what you capture vs. what you want to extract value wise).

Reply:
Good point.  Added this.